File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/play/clues.ts
/**
 * Campaign clues the DM may reveal to one character, the whole party, or
 * nobody (hidden). See shared.ts for the ownership model.
 */

import { getDb } from '../../db.ts';
import type { ApiResult, JsonValue } from '../../types.ts';
import { authenticate, isActor, isApiResult, findCampaign, requireParticipant, findMemberByCharacterId } from './shared.ts';

type ClueRow = {
  clue_id: string;
  text: string;
  audience: string;
  character_id: string | null;
};

function clueBody(row: ClueRow): JsonValue {
  const body: Record<string, unknown> = { clue_id: row.clue_id, text: row.text, audience: row.audience };
  if (row.character_id !== null) {
    body.character_id = row.character_id;
  }
  return body as JsonValue;
}

function nextClueSequence(db: ReturnType<typeof getDb>, campaignId: string): number {
  const row = db
    .prepare('SELECT COALESCE(MAX(sequence), 0) AS max_sequence FROM play_campaign_clues WHERE campaign_id = ?')
    .get(campaignId) as { max_sequence: number };
  return row.max_sequence + 1;
}

export function createClue(authHeader: string | undefined, campaignId: string, body: JsonValue): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  if (campaign.owner !== actor.username) {
    return { status: 403, body: { error: 'only the campaign dm may create clues' } };
  }

  const clueId = body.clue_id;
  if (typeof clueId !== 'string' || clueId.length === 0) {
    return { status: 400, body: { error: 'clue_id must be a non-empty string' } };
  }

  const text = body.text;
  if (typeof text !== 'string' || text.length === 0) {
    return { status: 400, body: { error: 'text must be a non-empty string' } };
  }

  const audience = body.audience;
  if (audience !== 'character' && audience !== 'party' && audience !== 'hidden') {
    return { status: 400, body: { error: "audience must be exactly 'character', 'party', or 'hidden'" } };
  }

  let characterId: string | null = null;
  if (audience === 'character') {
    const rawCharacterId = body.character_id;
    if (typeof rawCharacterId !== 'string' || rawCharacterId.length === 0) {
      return { status: 400, body: { error: 'character_id is required for character audience' } };
    }
    const member = findMemberByCharacterId(db, campaignId, rawCharacterId);
    if (isApiResult(member)) {
      return { status: 400, body: { error: 'character_id is not a known campaign member character' } };
    }
    characterId = rawCharacterId;
  } else if (body.character_id !== undefined) {
    return { status: 400, body: { error: 'character_id must be omitted for this audience' } };
  }

  const existing = db
    .prepare('SELECT clue_id FROM play_campaign_clues WHERE campaign_id = ? AND clue_id = ?')
    .get(campaignId, clueId);
  if (existing) {
    return { status: 409, body: { error: 'clue_id already exists in this campaign' } };
  }

  const sequence = nextClueSequence(db, campaignId);
  db.prepare(
    'INSERT INTO play_campaign_clues (campaign_id, sequence, clue_id, text, audience, character_id) VALUES (?, ?, ?, ?, ?, ?)',
  ).run(campaignId, sequence, clueId, text, audience, characterId);

  return { status: 201, body: clueBody({ clue_id: clueId, text, audience, character_id: characterId }) };
}

export function listClues(authHeader: string | undefined, campaignId: string): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const rows = db
    .prepare(
      'SELECT clue_id, text, audience, character_id FROM play_campaign_clues WHERE campaign_id = ? ORDER BY sequence ASC',
    )
    .all(campaignId) as ClueRow[];

  if (actor.username === campaign.owner) {
    return { status: 200, body: { clues: rows.map((row) => clueBody(row)) } };
  }

  const memberRow = db
    .prepare('SELECT character_id FROM play_campaign_members WHERE campaign_id = ? AND username = ?')
    .get(campaignId, actor.username) as { character_id: string } | undefined;
  const ownCharacterId = memberRow ? memberRow.character_id : null;

  const visible = rows.filter(
    (row) => row.audience === 'party' || (row.audience === 'character' && row.character_id === ownCharacterId),
  );

  return { status: 200, body: { clues: visible.map((row) => clueBody(row)) } };
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T072004Z_claude_claude-sonnet-5_ruby-sinatra/routes/play/clues.rb
# Campaign clues the DM may reveal to one character, the party, or nobody
# (hidden). Clue IDs are unique per campaign.

def play_clue_payload(clue)
  payload = { clue_id: clue['clue_id'], text: clue['text'], audience: clue['audience'] }
  payload[:character_id] = clue['character_id'] if clue['audience'] == 'character'
  payload
end

post '/v1/play/campaigns/:id/clues' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_owner!(campaign, user)

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  clue_id = body['clue_id']
  text = body['text']
  audience = body['audience']
  character_id = body['character_id']

  halt 400, { error: 'invalid clue_id' }.to_json unless clue_id.is_a?(String) && !clue_id.empty?
  halt 400, { error: 'invalid text' }.to_json unless text.is_a?(String) && !text.empty?
  halt 400, { error: 'invalid audience' }.to_json unless %w[character party hidden].include?(audience)

  if audience == 'character'
    halt 400, { error: 'character_id required' }.to_json unless character_id.is_a?(String) && !character_id.empty?
    halt 400, { error: 'unknown character' }.to_json unless db.execute(
      'SELECT 1 FROM play_campaign_members WHERE campaign_id = ? AND character_id = ?',
      [campaign['id'], character_id]
    ).first
  else
    halt 400, { error: 'character_id must be omitted' }.to_json unless character_id.nil?
  end

  existing = db.execute(
    'SELECT 1 FROM play_clues WHERE campaign_id = ? AND clue_id = ?',
    [campaign['id'], clue_id]
  ).first
  halt 409, { error: 'clue already exists' }.to_json if existing

  sequence = db.execute(
    'SELECT COALESCE(MAX(sequence), -1) + 1 AS seq FROM play_clues WHERE campaign_id = ?',
    [campaign['id']]
  ).first['seq']

  db.execute(
    'INSERT INTO play_clues (campaign_id, sequence, clue_id, text, audience, character_id) VALUES (?, ?, ?, ?, ?, ?)',
    [campaign['id'], sequence, clue_id, text, audience, audience == 'character' ? character_id : nil]
  )

  status 201
  payload = { clue_id: clue_id, text: text, audience: audience }
  payload[:character_id] = character_id if audience == 'character'
  payload.to_json
end

get '/v1/play/campaigns/:id/clues' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  is_owner = require_play_participant!(campaign, user, 'not a campaign member')

  clues = db.execute(
    'SELECT * FROM play_clues WHERE campaign_id = ? ORDER BY sequence ASC',
    [campaign['id']]
  )

  unless is_owner
    member = db.execute(
      'SELECT character_id FROM play_campaign_members WHERE campaign_id = ? AND username = ?',
      [campaign['id'], user['username']]
    ).first
    own_character_id = member && member['character_id']

    clues = clues.select do |clue|
      case clue['audience']
      when 'party' then true
      when 'character' then clue['character_id'] == own_character_id
      else false
      end
    end
  end

  { clues: clues.map { |clue| play_clue_payload(clue) } }.to_json
end
